### PR TITLE
Add BidstreamClient and SharingClient, deprecate UID2Client

### PR DESCRIPTION
**Pull request by caroline-ttd**
Introduce two new clients BidStreamClient and SharingClient and deprecate the existing UID2Client.

### CODE REVIEW

File: src/test/java/com/uid2/client/test/IntegrationExamples.java [hunk 1]
```diff
@@ -4,19 +4,92 @@
...
+                    Thread.sleep(3000);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        });
+        refreshThread.start();
+
+        for (int i = 0; i < 5; ++i) {
+            // Since this block will run initially, it will fail at the first time. Subsequently, upon refreshing, the call should succeed.
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/client/test/IntegrationExamples.java]: Sure, I added it.

File: src/test/java/com/uid2/client/test/IntegrationExamples.java [hunk 2]
```diff
@@ -4,19 +4,92 @@
...
     final String TEST_SECRET_KEY = System.getenv("UID2_SECRET_KEY");
     final String TEST_TOKEN = System.getenv("UID2_TEST_TOKEN");
+    final String TEST_DOMAIN = System.getenv("UID2_TEST_DOMAIN");
+
+    @Test // - insert your API key & test uid & test domain before enabling this test
+    public void ExampleBidStreamClient() {
+        BidstreamClient client = new BidstreamClient(TEST_ENDPOINT, TEST_API_KEY, TEST_SECRET_KEY);
+        RefreshResponse refreshResult = client.refresh();
+        if (!refreshResult.isSuccess()) {
+            System.out.println("Failed to refresh keys: " + refreshResult.getReason());
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/client/test/IntegrationExamples.java:23]: Thank you for catching this. I added it.

File: src/test/java/com/uid2/client/EncryptionTestsV4.java [hunk 3]
```diff
@@ -104,7 +92,12 @@ public void crossPlatformConsistencyCheck_Decrypt() throws Exception {
         assertEquals(EXAMPLE_UID, res.getUid());
     }
 
-    private static void validateAdvertisingToken(String advertisingTokenString, IdentityScope identityScope, IdentityType identityType) {
+    public static void validateAdvertisingToken(String advertisingTokenString, IdentityScope identityScope, IdentityType identityType, int tokenVersion) {
+        if (tokenVersion == 2) {
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/client/EncryptionTestsV4.java]: I created a new class called TokenVersionForTesting for the token version

File: src/main/java/com/uid2/client/TokenHelper.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        try {
+            return Uid2Encryption.decrypt(token, keyContainer, now, keyContainer.getIdentityScope(), domainNameFromBidRequest, clientType);
+        } catch (Exception e) {
+            return DecryptionResponse.makeError(DecryptionStatus.INVALID_PAYLOAD);
+        }
+    }
+
+    public EncryptionDataResponse encryptRawUidIntoToken(String rawUid, Instant now) {
+        KeyContainer keyContainer = this.container.get();
+        if(keyContainer == null) {
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/main/java/com/uid2/client/TokenHelper.java]: Changed all `if(` to `if (`

File: src/main/java/com/uid2/client/PrivacyBits.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package com.uid2.client;
+
+import java.util.BitSet;
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/main/java/com/uid2/client/PrivacyBits.java:3]: Sure, added it

File: src/main/java/com/uid2/client/BidstreamClient.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    DecryptionResponse decryptTokenIntoRawUid(String token, String domainNameFromBidRequest, Instant now) {
+        return tokenHelper.decrypt(token, now, domainNameFromBidRequest, ClientType.BIDSTREAM);
+    }
+
+    public RefreshResponse refresh() {
+        return tokenHelper.refresh("/v2/key/bidstream");
+    }
+
+    public RefreshResponse refreshJson(String json) {
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/main/java/com/uid2/client/BidstreamClient.java]: Yeah, I changed it.

File: src/main/java/com/uid2/client/DecryptionStatus.java [hunk 7]
```diff
@@ -44,4 +44,8 @@ public enum DecryptionStatus {
       * to decrypt a UID2 token. Ensure the factory class matches the token type you are decrypting.
       */
     INVALID_IDENTITY_SCOPE,
+     /**
+      * INVALID_TOKEN_LIFETIME: If token generated timestamp is longer than allow_clock_skew_seconds
```
[the remark this answers is not in the document]
Comment (jon8787) [reply src/main/java/com/uid2/client/DecryptionStatus.java]: was this changed?

File: src/test/java/com/uid2/client/EncryptionTestsV4.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jon8787) [reply src/test/java/com/uid2/client/EncryptionTestsV4.java]: hmm ok why do we do it differently than C# ?

[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/client/EncryptionTestsV4.java]: I feel they are the same. in the C#, "TokenVersion" class is also only used for testing. [here](https://github.com/IABTechLab/uid2-client-net/blob/main/test/UID2.Client.Test/builder/AdvertisingTokenBuilder.cs#L101)

File: src/main/java/com/uid2/client/DecryptionStatus.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/main/java/com/uid2/client/DecryptionStatus.java]: My bad, that makes sense.

File: src/test/java/com/uid2/client/BidstreamClientTests.java [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    private void decryptAndAssertSuccess(String token, TokenVersionForTesting tokenVersion) {
+        DecryptionResponse response = bidstreamClient.decryptTokenIntoRawUid(token, null);
+        assertSuccess(response, tokenVersion);
+    }
+
+    static void assertSuccess(DecryptionResponse response, TokenVersionForTesting tokenVersion) {
+        assertTrue(response.isSuccess());
+        assertEquals(EXAMPLE_UID, response.getUid());
+        assertEquals(tokenVersion.ordinal() + 2, response.getAdvertisingTokenVersion());
+        assertEquals(tokenVersion.ordinal() + 2, response.getAdvertisingTokenVersion());
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/client/BidstreamClientTests.java:283]: My bad, will remove one.

**Pull request merged**